### PR TITLE
Paul/add lifi

### PR DESCRIPTION
**Pull request by paullinator**
### CHANGELOG

Does this branch warrant an entry to the CHANGELOG?

- [ ] Yes
- [ ] No

### Dependencies

 none

### Description

 none

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1206578167289565

### CODE REVIEW

File: src/partners/lifi.ts [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    try {
+      const response = await retryFetch(url)
+      if (!response.ok) {
+        const text = await response.text()
+        throw new Error(text)
+      }
+      const jsonObj = await response.json()
+      const transferResults = asTransfersResult(jsonObj)
+      for (const tx of transferResults.transfers) {
+        const txts = tx.receiving.timestamp ?? tx.sending.timestamp ?? 0
```
Comment (samholmes) [inline src/partners/lifi.ts:117]: Add a comment explaining why this identifier is `txts`.

File: src/partners/lifi.ts [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        if (txts === 0) {
+          throw new Error('No timestamp')
+        }
+        const { isoDate, timestamp } = smartIsoDateFromTimestamp(txts)
+
+        // Use biggystri
```
Comment (samholmes) [inline src/partners/lifi.ts:123]: biggystring*

File: src/partners/lifi.ts [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+          ),
+          rawTx: tx
+        }
+        ssFormatTxs.push(ssTx)
+        if (ssTx.isoDate > latestIsoDate) {
+          latestIsoDate = ssTx.isoDate
+        }
+      }
+      const endDate = new Date(endTime)
+      startTime = endDate.getTime()
```
Comment (samholmes) [inline src/partners/lifi.ts:154]: `startTime = endTime`?

File: src/partners/lifi.ts [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+      datelog(
+        `Lifi endDate:${new Date(
+          endDate
+        ).toISOString()} latestIsoDate:${latestIsoDate}`
```
Comment (samholmes) [inline src/partners/lifi.ts:158]: Why new Date over a Date object?

File: src/partners/lifi.ts [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+        break
+      }
+      retry = 0
+    } catch (e) {
+      datelog(e)
+      // Retry a few times with time delay to prevent throttling
+      retry++
+      if (retry <= MAX_RETRIES) {
+        datelog(`Snoozing ${60 * retry}s`)
+        await snooze(61000 * retry)
```
Comment (samholmes) [inline src/partners/lifi.ts:170]: log doesn't appear to match snooze interval (unless I'm just bad at math).

File: src/partners/lifi.ts [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+  //   integrator: asString
+  // })
+})
+
+// Define the cleaner for the whole JSON
+const asTransfersResult = asObject({
+  transfers: asArray(asTransfer)
+})
+
+type Transfer = ReturnType<typeof asTransfer>
```
Comment (samholmes) [inline src/partners/lifi.ts:70]: Type never used.

**Review by samholmes — CHANGES_REQUESTED** [review]

File: src/partners/lifi.ts (hunk 6 above)
Comment (paullinator) [reply src/partners/lifi.ts:72]: Weird that lint didn't catch that. fixed

**Review by samholmes — APPROVED** [review]

**Pull request merged**